src/main/scala/com.snowplowanalytics.snowplow.analytics.scalasdk/Event.scala

```diff
@@ -0,0 +1,240 @@
+                 event_version:            Option[String],
+                 event_fingerprint:        Option[String],
+                 true_tstamp:              Option[Instant]) {
+
+  def inventory: Set[Data.ShreddedType] = {
+    val unstructEvent = unstruct_event
+      .data
+      .toSet
+      .map((ue: SelfDescribingData[Json]) => Data.ShreddedType(Data.UnstructEvent, ue.schema))
+
+    val derivedContexts = derived_contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.DerivedContexts), ctx.schema))
+
+    val customContexts = contexts
+      .data
+      .toSet
+      .map((ctx: SelfDescribingData[Json]) => Data.ShreddedType(Data.Contexts(Data.CustomContexts), ctx.schema))
+
+    customContexts ++ derivedContexts ++ unstructEvent
+  }
+}
+
+
+object Event {
+  case class Lossy(event: Event) extends AnyVal
+  case class Lossless(event: Event) extends AnyVal
+
+  implicit val jsonEncoder: Encoder[Event] = deriveEncoder[Event]
+  implicit val lossyEncoder: Encoder[Lossy] = Encoder.instance { lossy: Lossy =>
+    // Convert event to initial JSON
+    val event = lossy.event
+    val json = event.asJson.asObject.getOrElse(JsonObject()).toList
+
+    // Filter out autoderived contexts and transform unstruct_event
+    val transformed =
+      json
+        .patch(json.indexWhere(_._1 == "unstruct_event"), transformUnstruct(event.unstruct_event).toSeq, 1)
+        .filterNot(_._1 == "contexts")
+        .filterNot(_._1 == "derived_contexts")
+
+    // Create geo_location from latitude/longitude
+    val latitude = json.find(_._1 == "geo_latitude")
+    val longitude = json.find(_._1 == "geo_longitude")
+    val geoLocation = if (latitude.isDefined && longitude.isDefined) s"${latitude.get._2},${longitude.get._2}" else ""
+
+    // Add lossy contexts
+    val transformedWithContexts =
+      if (geoLocation.isEmpty)
+        transformed ++ transformContexts(event.derived_contexts) ++ transformContexts(event.contexts)
+      else
+        ("geo_location", geoLocation.asJson) :: (transformed ++ transformContexts(event.derived_contexts) ++ transformContexts(event.contexts))
+
+    // Convert back to JSON
+    JsonObject(transformedWithContexts.reverse: _*).asJson
+  }
+  private val parser: Parser[Event] = Parser.deriveFor[Event].get
+
+  def parse(line: String): RowDecodeResult[Event] =
+    parser.parse(line)
+
+  def asJson(event: Event, lossy: Boolean): Json =
+    if (lossy) {
+      Lossy(event).asJson
+    } else {
+      event.asJson
+    }
+
+  def transformSchema(prefix: String, vendor: String, name: String, model: Int): String = {
+    val snakeCaseVendor = vendor.replaceAll("""[\.\-]""", "_").toLowerCase
+    val snakeCaseName = name.replaceAll("""[\.\-]""", "_").replaceAll("([^A-Z_])([A-Z])", "$1_$2").toLowerCase
+    s"${prefix}_${snakeCaseVendor}_${snakeCaseName}_$model"
+  }
+
+  def transformContexts(c: Contexts) = {
```

Comment:
Why these are in `Event` companion object, not in corresponding `Contexts`/`UnstructEvent`?